### PR TITLE
#677 add pattern Trampoline.

### PR DESCRIPTION
**Pull request by besok**
Pull request title

- Add pattern trampoline.
- #677 -> https://github.com/iluwatar/java-design-patterns/issues/677

Pull request description

- Add readme, code,test.

> For detailed contributing instructions see https://github.com/iluwatar/java-design-patterns/wiki/01.-How-to-contribute

### CODE REVIEW

**Comment by besok** [comment]
Throttling test has failed. 
I don't understand what i should do now.

**Comment by iluwatar** [comment]
It is a known issue https://github.com/iluwatar/java-design-patterns/issues/643

File: trampoline/pom.xml [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            <scope>test</scope>
+        </dependency>
+        <!-- https://mvnrepository.com/artifact/org.projectlombok/lombok -->
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <version>1.16.18</version>
+            <scope>provided</scope>
+        </dependency>
+
```
Comment (iluwatar) [inline trampoline/pom.xml]: Using Lombok should be project level decision. We haven't used it elsewhere so please do not take it into use in one pattern.

File: trampoline/src/main/java/com/iluwatar/trampoline/Trampoline.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+import java.util.stream.Stream;
+
+/**When get is called on the returned Trampoline, internally it will iterate calling ‘jump’
+ on the returned Trampoline as long as the concrete instance returned is More,
+ stopping once the returned instance is Done. Essential we convert looping via recursion into iteration,
+ the key enabling mechanism is the fact that Trampoline.more is a lazy operation.
+ Trampoline in cyclops-react extends java.util.Supplier. Calling Trampoline.more we are basically creating
+ a Supplier that defers the actual recursive call, and having defered the call we can move it outside of the recursive loop.
+ This means we can define algorithms recursively in Java but execute them iteratively.*/
```
Comment (iluwatar) [inline trampoline/src/main/java/com/iluwatar/trampoline/Trampoline.java]: Please refer to methods properly in Javadoc e.g. {@link package.class#member label}. What is cyclops-react?

File: trampoline/src/main/java/com/iluwatar/trampoline/TrampolineApp.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ Trampoline has 2 state : [done], [ more]
+ </div>
+ When get is called on the returned Trampoline, internally it will iterate calling ‘jump’
+ on the returned Trampoline as long as the concrete instance returned is More,
+ stopping once the returned instance is Done. Essential we convert looping via recursion into iteration,
+ the key enabling mechanism is the fact that Trampoline.more is a lazy operation.
+ Trampoline in cyclops-react extends java.util.Supplier. Calling Trampoline.more we are basically creating
+ a Supplier that defers the actual recursive call, and having defered the call we can move it outside of the recursive loop.
+ This means we can define algorithms recursively in Java but execute them iteratively.
+ */
```
Comment (iluwatar) [inline trampoline/src/main/java/com/iluwatar/trampoline/TrampolineApp.java]: Please prefer \<p\> tag in Javadoc. See other patterns for examples. Please describe what the example does and how it implements the pattern.

File: trampoline/src/test/java/com/iluwatar/trampoline/TrampolineAppTest.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+import org.junit.Test;
+
+import java.io.IOException;
+
+import static org.junit.Assert.*;
+
+public class TrampolineAppTest {
+
+    @Test
+    public void test()throws IOException{
```
Comment (iluwatar) [inline trampoline/src/test/java/com/iluwatar/trampoline/TrampolineAppTest.java]: Please use descriptive test method name? What does it test?

**Comment by iluwatar** [comment]
@besok please comment when you've implemented the changes.

**Comment by besok** [comment]
Ok. I am going to do it on holidays(20-21.01)

File: trampoline/src/test/java/com/iluwatar/trampoline/TrampolineAppTest.java (hunk 4 above)
Comment (besok) [reply trampoline/src/test/java/com/iluwatar/trampoline/TrampolineAppTest.java]: Done. Changed to testTrampolineWithFactorialFunction

File: trampoline/src/main/java/com/iluwatar/trampoline/TrampolineApp.java (hunk 3 above)
Comment (besok) [reply trampoline/src/main/java/com/iluwatar/trampoline/TrampolineApp.java]: done

File: trampoline/src/main/java/com/iluwatar/trampoline/Trampoline.java (hunk 2 above)
Comment (besok) [reply trampoline/src/main/java/com/iluwatar/trampoline/Trampoline.java]: cyclops-react is a library for additions java8 streams.
done.

File: trampoline/pom.xml (hunk 1 above)
Comment (besok) [reply trampoline/pom.xml]: done

File: trampoline/README.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+---
+
+## Intent
+By representing a computation in one of 2 states 
+(completed with result, or a reference to the reminder of the computation, 
+something like the way a java.util.Supplier does) 
+it is possible to implement algorithms recursively in Java without blowing the stack 
+and to interleave the execution of functions without hard coding them together or even using threads.
+
+
```
[the remark this answers is not in the document]
Comment (besok) [reply trampoline/README.md]: i added section,

File: trampoline/README.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+The time overhead of a reflection call is traded for the space overhead of an inner class. 
+Trampolines in Java usually involve the creation of a GenericListener to pass events to an outer class.
+
+
+## Credits
+
+* [Trampolining: a practical guide for awesome Java Developers](https://medium.com/@johnmcclean/trampolining-a-practical-guide-for-awesome-java-developers-4b657d9c3076)
+* [Trampoline in java ](http://mindprod.com/jgloss/trampoline.html)
+* [cyclops-react](https://github.com/aol/cyclops-react)
+
```
[the remark this answers is not in the document]
Comment (besok) [reply trampoline/README.md]: changed it.

**Comment by besok** [comment]
> The project is missing from parent pom.xml

How can i add this modula to parent pom?
I added to module section 
` </modules>
	...
          <module>trampoline</module>
    </modules>
`

**Comment by besok** [comment]
@iluwatar , can you check my changes?

**Comment by iluwatar** [comment]
Well done @besok :+1: Thanks for the new pattern!

**Pull request merged**